### PR TITLE
Lux 

### PR DESCRIPTION
**Pull request by AlCap23**
- [x] RandomSearch
- [x] OccamNet
- [ ] DeepSymbolicRegression

~~Probably also Bayesian and GP~~

It would be awesome to provide a BinaryTree container along the LayeredDAG, but I think we can do a PR afterwards.

Also : CrossEntropyMethod and maybe an API for POMPDS / ReinforcementLearning ?

### CODE REVIEW

File: lib/DataDrivenLux/src/DataDrivenLux.jl [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+using DataDrivenDiffEq.DocStringExtensions
+using DataDrivenDiffEq.CommonSolve
+using DataDrivenDiffEq.CommonSolve: solve!
+using DataDrivenDiffEq.StatsBase
+using DataDrivenDiffEq.Parameters
+using DataDrivenDiffEq.Setfield
+
+using Reexport
+@reexport using Optim
```
Comment (ChrisRackauckas) [inline lib/DataDrivenLux/src/DataDrivenLux.jl:22]: Not using Optimization? Pure BFGS would be prone to local minima?

File: lib/DataDrivenLux/src/algorithms/reinforce.jl [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    skip::Bool = true
+    "Evaluation function to sort the samples"
+    loss::L = aicc
+    "The number of candidates to keep in each iteration"
+    keep::Union{Real, Int} = 0.1
+    "Use protected operators"
+    use_protected::Bool = true
+    "Use distributed optimization and resampling"
+    distributed::Bool = false
+    "Use threaded optimization and resampling - not implemented right now."
```
Comment (ChrisRackauckas) [inline lib/DataDrivenLux/src/algorithms/reinforce.jl:68]: would it ever be smart? If you're already using BLAS, you're already multithreading

File: lib/DataDrivenLux/src/lux/layer.jl [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+Base.getindex(c::FunctionLayer, i::Int) = c.nodes[i]
+
+Base.length(c::FunctionLayer) = length(c.nodes)
+Base.lastindex(c::FunctionLayer) = lastindex(c.nodes)
+Base.firstindex(c::FunctionLayer) = firstindex(c.nodes)
+
+function get_loglikelihood(r::FunctionLayer, ps, st)
+    _get_layer_loglikelihood(r.nodes, ps, st)
+end
+
```
Comment (ChrisRackauckas) [inline lib/DataDrivenLux/src/lux/layer.jl:53]: Compile times will be fun, but we can worry about that later.

**Comment by ChrisRackauckas** [comment]
This looks reasonable enough. A few small comments but nothing major. This could just merge as-is.

File: lib/DataDrivenLux/src/algorithms/reinforce.jl (hunk 2 above)
Comment (AlCap23) [reply lib/DataDrivenLux/src/algorithms/reinforce.jl:68]: I've had a discussion on that with @mohamed82008 and @korsbo . Seems to work fine.

**Comment by AlCap23** [comment]
I'll add a few more tests for Gumbel softmax and the deep symbolic regression stuff and merge. 

Docs will happen soonish afterwards.

File: lib/DataDrivenLux/src/DataDrivenLux.jl (hunk 1 above)
Comment (AlCap23) [reply lib/DataDrivenLux/src/DataDrivenLux.jl:22]: I'll benchmark later today how this would affect the performance of the optimization routine. 👍

Comment (AlCap23) [reply lib/DataDrivenLux/src/DataDrivenLux.jl:22]: I've tested Optimization and the overhead seems a little too much imho


```Julia

fs = (exp,)
arities = (1,)
dag = LayeredDAG(1, 1, 1, arities, fs, skip = true)
X = permutedims(collect(0:0.1:3.0))
Y = sin.(2.0 * X)
@variables x
@parameters p [bounds = (1.0, 2.5), dist = Normal(1.75, 1.0)]
basis = Basis([sin(p * x)], [x], parameters = [p])

dataset = Dataset(X, Y)
rng = StableRNG(2)
candidate = DataDrivenLux.Candidate(rng, dag, basis, dataset)


@btime DataDrivenLux.optimize_candidate!($candidate, $dataset)
# Optim
# 138.708 μs (3231 allocations: 204.42 KiB)
# Optimization
# 1.244 ms (28232 allocations: 2.44 MiB)
```

This constructs the OptimizationFunction once and remakes the problem in the optimize call with initial parameters and the current configuration. Given that this needs to happen for each iteration a few hundred of times, this would slow down the algorithm massively.
